rustler_bigint/src/big_int.rs

```diff
@@ -0,0 +1,396 @@
+/// assert_eq!(number, bigint);
+/// ```
+///
+/// ## Examples
+///
+/// ```rust
+/// use rustler_bigint::BigInt;
+///
+/// #[rustler::nif]
+/// pub fn pow(base: BigInt, exponent: u32) -> BigInt {
+///   base.pow(exponent).into()
+/// }
+/// ```
+///
+/// ```rust
+/// use rustler::Binary;
+/// use rustler_bigint::{BigInt, num_bigint};
+///
+/// #[rustler::nif]
+/// pub fn binary_to_integer(binary: Binary) -> BigInt {
+///   num_bigint::BigInt::from_signed_bytes_be(binary.as_slice()).into()
+/// }
+/// ```
+///
+/// ```rust
+/// use rustler::{Binary, Env, NewBinary};
+/// use rustler_bigint::BigInt;
+///
+/// #[rustler::nif]
+/// pub fn integer_to_binary<'a>(env: Env<'a>, integer: BigInt) -> Binary<'a> {
+///   let bytes = integer.to_signed_bytes_be();
+///   let mut output = NewBinary::new(env, bytes.len());
+///   output.as_mut_slice().copy_from_slice(bytes.as_slice());
+///   output.into()
+/// }
+/// ```
+///
+#[derive(Debug, PartialEq, PartialOrd)]
+pub struct BigInt(num_bigint::BigInt);
+
+impl std::convert::From<num_bigint::BigInt> for BigInt {
+    fn from(big_int: num_bigint::BigInt) -> BigInt {
+        BigInt(big_int)
+    }
+}
+
+impl std::convert::From<BigInt> for num_bigint::BigInt {
+    fn from(big_int: BigInt) -> num_bigint::BigInt {
+        big_int.0
+    }
+}
+
+impl std::ops::Deref for BigInt {
+    type Target = num_bigint::BigInt;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl std::ops::DerefMut for BigInt {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+fn decode_big_integer(input: &[u8]) -> NifResult<num_bigint::BigInt> {
+    if input[0] != 131 {
+        return Err(Error::BadArg);
+    };
+
+    let big_int = match input[1] {
+        97 => {
+            // small integer / byte
+            num_bigint::BigInt::from(input[2])
+        }
+
+        98 => {
+            // integer
+            num_bigint::BigInt::from_signed_bytes_be(&input[2..6])
+        }
+
+        110 => {
+            // small big integer
+            let n = input[2] as usize;
+            let sign = if input[3] == 0 {
+                Sign::Plus
+            } else {
+                Sign::Minus
+            };
+
+            num_bigint::BigInt::from_bytes_le(sign, &input[4..n + 4])
+        }
+
+        111 => {
+            // large big integer
+            let n = u32::from_be_bytes([input[2], input[3], input[4], input[5]]) as usize;
+            let sign = if input[6] == 0 {
+                Sign::Plus
+            } else {
+                Sign::Minus
+            };
+
+            num_bigint::BigInt::from_bytes_le(sign, &input[7..n + 7])
+        }
+
+        _ => return Err(Error::BadArg),
+    };
+
+    Ok(big_int)
+}
+
+fn encode_big_integer(big_int: &num_bigint::BigInt) -> Vec<u8> {
+    let (sign, data) = big_int.to_bytes_le();
+    let sign = if sign == Sign::Minus { 1 } else { 0 };
+
+    let mut out = vec![131];
+    if data.len() < 256 {
```

Comment:
Should we also encode into small integer and integer?

Reply:
I thought about it, but we load it into erlang in the next step. And i thought maybe the logic to implement this is "more work" than the runtime overhead it takes to load a bit larger number.

I shall write a small benchmark to verify my assumptions

Reply:
Written a small benchee benchmark, which loads the number 50 if big integer format and in small integer format
```
big_int = <<131, 110, 1, 0, 50>>
small_int = <<131, 97, 50>>
int = <<131, 98, 0, 0, 0, 50>>

Benchee.run(
  %{
    "small_int" => fn -> :erlang.binary_to_term(small_int) end,
    "int" => fn -> :erlang.binary_to_term(int) end,
    "big_int" => fn -> :erlang.binary_to_term(big_int) end
  }
)

```

out:

```
Comparison:
int             15.64 M
small_int       15.21 M - 1.03x slower +1.81 ns
big_int         14.22 M - 1.10x slower +6.36 ns
```

big int is indeed a bit slower 🤔

Reply:
good news, found an easy way to implement it, so I have added decoding as normal integer (98)

Reply:
>  ran it a couple of times and the results are all over the place :D so i dont think there is a slow down between the three

Yeah, for such benchmarks, Erlang's scheduling will often result in flaky results. You have to re-run within the benchmark itself using a loop and then adjust the results, that often helps.